triedb/firewood/database.go

```diff
@@ -0,0 +1,610 @@
+type ProposalContext struct {
+	Proposal *ffi.Proposal
+	Hashes   map[common.Hash]struct{} // All corresponding block hashes
+	Root     common.Hash
+	Block    uint64
+	Parent   *ProposalContext
+	Children []*ProposalContext
+}
+
+type Config struct {
+	FilePath             string
+	CleanCacheSize       int  // Size of the clean cache in bytes
+	FreeListCacheEntries uint // Number of free list entries to cache
+	Revisions            uint
+	ReadCacheStrategy    ffi.CacheStrategy
+}
+
+// Note that `FilePath` is not specificied, and must always be set by the user.
+var Defaults = &Config{
+	CleanCacheSize:       1024 * 1024, // 1MB
+	FreeListCacheEntries: 40_000,
+	Revisions:            100,
+	ReadCacheStrategy:    ffi.CacheAllReads,
+}
+
+func (c Config) BackendConstructor(_ ethdb.Database) triedb.DBOverride {
+	return New(&c)
+}
+
+type Database struct {
+	fwDisk       *ffi.Database // The underlying Firewood database, used for storing proposals and revisions.
+	proposalLock sync.RWMutex
+	// proposalMap provides O(1) access by state root to all proposals stored in the proposalTree
+	proposalMap map[common.Hash][]*ProposalContext
+	// The proposal tree tracks the structure of the current proposals, and which proposals are children of which.
+	// This is used to ensure that we can dereference proposals correctly and commit the correct ones
+	// in the case of duplicate state roots.
+	// The root of the tree is stored here, and represents the top-most layer on disk.
+	proposalTree *ProposalContext
+}
+
+// New creates a new Firewood database with the given disk database and configuration.
+// Any error during creation will cause the program to exit.
+func New(config *Config) *Database {
+	if config == nil {
+		config = Defaults
+	}
+
+	fwConfig, err := validatePath(config)
+	if err != nil {
+		log.Crit("firewood: error validating config", "error", err)
+	}
+
+	fw, err := ffi.New(config.FilePath, fwConfig)
+	if err != nil {
+		log.Crit("firewood: error creating firewood database", "error", err)
+	}
+
+	currentRoot, err := fw.Root()
+	if err != nil {
+		log.Crit("firewood: error getting current root", "error", err)
+	}
+
+	return &Database{
+		fwDisk:      fw,
+		proposalMap: make(map[common.Hash][]*ProposalContext),
+		proposalTree: &ProposalContext{
+			Root: common.Hash(currentRoot),
+		},
+	}
+}
+
+func validatePath(trieConfig *Config) (*ffi.Config, error) {
+	if trieConfig.FilePath == "" {
+		return nil, fmt.Errorf("firewood database file path must be set")
+	}
+
+	// Check that the directory exists
+	dir := filepath.Dir(trieConfig.FilePath)
+	info, err := os.Stat(dir)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, fmt.Errorf("error checking database directory: %w", err)
+		}
+		log.Info("Database directory not found, creating", "path", dir)
+		if err := os.MkdirAll(dir, 0755); err != nil {
+			return nil, fmt.Errorf("error creating database directory: %w", err)
+		}
+	}
```

Comment:
just bc it's hard to follow `!os.IsNotExist` (double negative)
```suggestion
      	if os.IsNotExist(err) {
      		log.Info("Database directory not found, creating", "path", dir)
      		if err := os.MkdirAll(dir, 0755); err != nil {
      			return nil, fmt.Errorf("error creating database directory: %w", err)
      		}
      	} else {
      		return nil, fmt.Errorf("error checking database directory: %w", err)
      	}
```

Reply:
This wasn't quite right (didn't handle `err==nil`) so I fixed it locally